File: Leetcode/0099-recover-binary-search-tree/solution.go
package main

import "fmt"

type TreeNode struct {
	Val         int
	Left, Right *TreeNode
}

func recoverTree(root *TreeNode) {
	var first, second, prev *TreeNode
	var inorder func(n *TreeNode)
	inorder = func(n *TreeNode) {
		if n == nil {
			return
		}
		inorder(n.Left)
		if prev != nil && prev.Val > n.Val {
			if first == nil {
				first = prev
			}
			second = n
		}
		prev = n
		inorder(n.Right)
	}
	inorder(root)
	first.Val, second.Val = second.Val, first.Val
}

func buildTree(arr []interface{}) *TreeNode {
	if len(arr) == 0 || arr[0] == nil {
		return nil
	}
	root := &TreeNode{Val: arr[0].(int)}
	q := []*TreeNode{root}
	i := 1
	for len(q) > 0 && i < len(arr) {
		n := q[0]
		q = q[1:]
		if i < len(arr) && arr[i] != nil {
			n.Left = &TreeNode{Val: arr[i].(int)}
			q = append(q, n.Left)
		}
		i++
		if i < len(arr) && arr[i] != nil {
			n.Right = &TreeNode{Val: arr[i].(int)}
			q = append(q, n.Right)
		}
		i++
	}
	return root
}

func inorder(n *TreeNode) []int {
	if n == nil {
		return nil
	}
	r := inorder(n.Left)
	r = append(r, n.Val)
	r = append(r, inorder(n.Right)...)
	return r
}

func isSorted(a []int) bool {
	for i := 1; i < len(a); i++ {
		if a[i] < a[i-1] {
			return false
		}
	}
	return true
}

func main() {
	passed, failed := 0, 0
	test := func(name string, root *TreeNode) {
		recoverTree(root)
		if isSorted(inorder(root)) {
			fmt.Printf("PASS: %s → %v\n", name, inorder(root))
			passed++
		} else {
			fmt.Printf("FAIL: %s → %v\n", name, inorder(root))
			failed++
		}
	}
	cases := []struct {
		name string
		arr  []interface{}
	}{
		{"Example 1", []interface{}{1, 3, nil, nil, 2}},
		{"Example 2", []interface{}{3, 1, 4, nil, nil, 2}},
		{"Adjacent swap", []interface{}{1, 2}},
	}
	for _, c := range cases {
		test(c.name, buildTree(c.arr))
	}
	fmt.Printf("\nResults: %d passed, %d failed, %d total\n", passed, failed, passed+failed)
}

File: Leetcode/0099-recover-binary-search-tree/solution.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val; self.left = left; self.right = right


class Solution:
    def recoverTree(self, root: Optional[TreeNode]) -> None:
        """Time O(n), Space O(h)."""
        self.first = self.second = self.prev = None
        def inorder(n):
            if not n: return
            inorder(n.left)
            if self.prev and self.prev.val > n.val:
                if not self.first: self.first = self.prev
                self.second = n
            self.prev = n
            inorder(n.right)
        inorder(root)
        self.first.val, self.second.val = self.second.val, self.first.val


def build_tree(arr):
    if not arr or arr[0] is None: return None
    root = TreeNode(arr[0])
    q = [root]; i = 1
    while q and i < len(arr):
        n = q.pop(0)
        if i < len(arr) and arr[i] is not None: n.left = TreeNode(arr[i]); q.append(n.left)
        i += 1
        if i < len(arr) and arr[i] is not None: n.right = TreeNode(arr[i]); q.append(n.right)
        i += 1
    return root


def inorder(n):
    if not n: return []
    return inorder(n.left) + [n.val] + inorder(n.right)


if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0
    cases = [
        ("Example 1", [1, 3, None, None, 2]),
        ("Example 2", [3, 1, 4, None, None, 2]),
        ("Adjacent swap", [1, 2]),
    ]
    for name, t in cases:
        root = build_tree(t)
        Solution().recoverTree(root)
        r = inorder(root)
        if all(r[i] >= r[i - 1] for i in range(1, len(r))):
            print(f"PASS: {name} → {r}"); passed += 1
        else:
            print(f"FAIL: {name} → {r}"); failed += 1
    print(f"\nResults: {passed} passed, {failed} failed, {passed + failed} total")
